### PR TITLE
implements BridgeCluster controller and friends

### PR DESCRIPTION
**Pull request by wilybrace**
- Adds CRD validation constraints for CrunchyBridge types
 - Updates example CR for BridgeCluster
 - Adds conditional compilation for DBaaS support
 - Adds interface for unsetting credential provider
 - Adds interface for invalidating client API session
 - Removes deprecated InstanceDetail from ClusterDetail

`bridgeapi.UnsetLogin` wipes the underlying `CredentialProvider` from the state - only necessary if you have a static credential source.
`bridgeapi.Logout` resets the login state of the client library, allowing it to perform a new login on-demand using credentials from the credential provider. This is useful if you've already updated the kubernetes secret holding the credentials to new values and you want to force forgetting the old values without providing a new `CredentialProvider`

### CODE REVIEW

File: controllers/crunchybridge/bridgecluster_controller.go [hunk 1]
```diff
@@ -39,17 +53,121 @@ type BridgeClusterReconciler struct {
...
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
 func (r *BridgeClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx, "BridgeCluster", req.NamespacedName)
+
+	if r.BridgeClient == nil {
```
Comment (redhatHameed) [inline controllers/crunchybridge/bridgecluster_controller.go:62]: where we are setting API credentials for cluster creation?

File: controllers/crunchybridge/bridgecluster_controller.go [hunk 2]
```diff
@@ -39,17 +53,121 @@ type BridgeClusterReconciler struct {
...
+			}
+			controllerutil.RemoveFinalizer(clusterObj, bcFinalizer)
+			if err := r.Update(ctx, clusterObj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		switch clusterObj.Status.Phase {
+		case crunchybridgev1alpha1.PhaseUnknown:
+			fmt.Printf("Phase: Unknown\n")
```
Comment (redhatHameed) [inline controllers/crunchybridge/bridgecluster_controller.go:94]: can we logger here?

File: controllers/crunchybridge/bridgecluster_controller.go [hunk 3]
```diff
@@ -39,17 +53,121 @@ type BridgeClusterReconciler struct {
...
+
+			// Set pending phase after so any errors in setting finalizer
+			// don't advance state
+			clusterObj.Status.Phase = crunchybridgev1alpha1.PhasePending
+			if err := r.Status().Update(ctx, clusterObj); err != nil {
+				return ctrl.Result{}, err
+			}
+
+		case crunchybridgev1alpha1.PhasePending:
+			fmt.Printf("Phase: Pending\n")
```
Comment (redhatHameed) [inline controllers/crunchybridge/bridgecluster_controller.go:111]: same here can we use logger?

File: controllers/crunchybridge/bridgecluster_controller.go [hunk 4]
```diff
@@ -39,17 +53,121 @@ type BridgeClusterReconciler struct {
...
+
+			// Assuming the request was sent, update phase
+			clusterObj.Status.Phase = crunchybridgev1alpha1.PhaseCreating
+			clusterObj.Status.Updated = time.Now().Format(time.RFC3339)
+			if err := r.Status().Update(ctx, clusterObj); err != nil {
+				return ctrl.Result{}, err
+			}
+
+		case crunchybridgev1alpha1.PhaseCreating:
+			fmt.Printf("Phase: %s - ver: %s\n", clusterObj.Status.Phase, clusterObj.ResourceVersion)
```
Comment (redhatHameed) [inline controllers/crunchybridge/bridgecluster_controller.go:129]: replace with logger?

File: dbaas_support.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
+// +build dbaas
```
Comment (redhatHameed) [inline dbaas_support.go:1]: why moved from main. go as we cluster creation also will be part DBaaS future requirements?

File: main.go [hunk 6]
```diff
@@ -48,17 +53,27 @@ func init() {
...
-	var metricsAddr string
+	// Variables from boilerplate
+	var metricsAddr, probeAddr string
 	var enableLeaderElection bool
-	var probeAddr string
+
 	var crunchybridgeAPIURL string
-	flag.StringVar(&crunchybridgeAPIURL, "crunchybridgeapi-url", "https://api.crunchybridge.com", "the Crunchy bridge API URL (no slash in the end).")
+	// Namespace and Name for APIKey secret default values
+	credNamespace := "default"
```
Comment (redhatHameed) [inline main.go:73]: why adding default namespace for credentials? we can pass in CR reference as done in inventory CR?

File: main.go [hunk 7]
```diff
@@ -48,17 +53,27 @@ func init() {
...
+	credName := "crunchybridge_api_key"
```
Comment (redhatHameed) [inline main.go:74]: same we can put into CR?

**Review by redhatHameed — COMMENTED** [review]
@wilybrace added few comments, Further going to verify this on OCP, and will update you thanks

File: Makefile [hunk 8]
```diff
@@ -101,11 +104,11 @@ test: manifests generate fmt vet ## Run tests.
...
 ##@ Build
 
-build: generate fmt vet ## Build manager binary.
-	go build -o bin/manager main.go
+build: generate fmt vet ## Build manager binary, do not explicitly list files so build flags do their job.
+	go build -tags '${GOTAGS}' -o bin/manager
 
 run: manifests generate fmt vet ## Run a controller from your host.
-	go run ./main.go
+	go run ./main.go ./dbaas_support.go
```
[the remark this answers is not in the document]
Comment (wilybrace) [reply Makefile:111]: This is part of the conditional compilation for DBaaS support, the manager hooks for DBaaS controllers are in dbaas_support and included when the build tag `dbaas` is used.

File: controllers/crunchybridge/bridgecluster_controller.go (hunk 1 above)
Comment (wilybrace) [reply controllers/crunchybridge/bridgecluster_controller.go:62]: The reconciler is given a client when initialized in `main` - either a client using configured credentials from the dbaas setup, or a new client configured from being pointed to a secret object identified by environment variables.

File: controllers/crunchybridge/bridgecluster_controller.go (hunk 2 above)
Comment (wilybrace) [reply controllers/crunchybridge/bridgecluster_controller.go:94]: The non-error `fmt` calls were supposed to be swapped with logger calls, will remove/fix.

File: dbaas_support.go (hunk 5 above)
Comment (wilybrace) [reply dbaas_support.go:1]: As mentioned in a previous comment, this was to support running the operator in non-openshift environments where the dbaas support is not required for the operator and the additional controllers are not required.

File: main.go (hunk 6 above)
Comment (wilybrace) [reply main.go:73]: The credentials are wherever they are created, this is a variable that holds the environment-configured location (namespace and name) of where the credentials live. Using a CR to bootstrap the DBaaS system is a design decision for the DBaaS approach to providers but is not yet necessary for a non-DBaaS operator.

File: main.go (hunk 7 above)
Comment (wilybrace) [reply main.go:74]: These variables are not used when DBaaS support is enabled, they are only for 'standalone operator' use.

File: main.go (hunk 6 above)
Comment (redhatHameed) [reply main.go:73]: understood thanks, can be added these environment variables for a container like https://github.com/CrunchyData/crunchy-bridge-operator/blob/master/config/manager/manager.yaml#L56 WDYT?

File: controllers/crunchybridge/bridgecluster_controller.go (hunk 1 above)
Comment (redhatHameed) [reply controllers/crunchybridge/bridgecluster_controller.go:62]: can we add these few lines in the readme file about configuring the credentails option?

Comment (wilybrace) [reply controllers/crunchybridge/bridgecluster_controller.go:62]: Agreed, that makes a lot of sense - good catch

File: internal/bridgeapi/login.go [hunk 9]
```diff
@@ -190,7 +230,23 @@ func SetLogin(cp CredentialProvider, authBaseURL *url.URL) {
 	}
 }
 
+// Resets the LoginManager state to nearly new, ready for a new SetLogin call
+func UnsetLogin() {
```
Comment (redhatHameed) [inline internal/bridgeapi/login.go:234]: `UnsetLogin` and `Logout` - which function is more suitable for the DBaaS controllers for invalidating tokens.

Comment (wilybrace) [reply internal/bridgeapi/login.go:234]: _it _depends__

Both invalidate the current login state by forgetting the current token. However, `UnsetLogin` resets the CredentialProvider, which means `SetLogin` *must* be called with valid credentials before anything will work again.
`Logout` leaves the credential provider in place, which will be used on the next call to any of the client data functions. That is, any attempt to contact the API at that point will login with the credentials from the provider, which would be the kube secret specified for the credential provider.

File: Makefile (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeremyary) [reply Makefile:111]: I'm digging the build tag pattern, will have to remember this one :)

File: controllers/crunchybridge/bridgecluster_controller.go [hunk 10]
```diff
@@ -39,17 +53,117 @@ type BridgeClusterReconciler struct {
...
+				}
+			}
+			controllerutil.RemoveFinalizer(clusterObj, bcFinalizer)
+			if err := r.Update(ctx, clusterObj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		switch clusterObj.Status.Phase {
+		case crunchybridgev1alpha1.PhaseUnknown:
```
Comment (jeremyary) [inline controllers/crunchybridge/bridgecluster_controller.go:94]: Switch/case format in the reconcile's a nice touch - I feel a nice visual break between the responsibilities and easily identifiable parts for future reference :clap:

**Review by jeremyary — APPROVED** [review]

**Pull request merged**